src/sqlfluff/core/rules/analysis/select_crawler.py

```diff
@@ -1,126 +1,43 @@
 """Tools for more complex analysis of SELECT statements."""
-from collections import defaultdict
+from dataclasses import dataclass, field
+from enum import Enum
 from typing import Dict, Generator, List, NamedTuple, Optional, Union
 
 from cached_property import cached_property
 
 from sqlfluff.core.dialects.common import AliasInfo
 from sqlfluff.core.dialects.base import Dialect
-from sqlfluff.core.parser.segments.base import BaseSegment
+from sqlfluff.core.parser import BaseSegment
 from sqlfluff.core.rules.analysis.select import get_select_statement_info
 
 
+class QueryType(Enum):
```

Comment:
This module has been heavily reworked, not only to support nested CTEs, but also:
* Improved object model that matches the structure of SQL better
* Streamlined the API of `SelectCrawler`